tests/scenario/test_charm.py

```diff
@@ -0,0 +1,601 @@
+                return_code=0, stdout=version
+            )
+        },
+    )
+    storage = scenario.Storage("local-media")
+    peer_relation = scenario.PeerRelation(
+        "takahe-peer", "takahe-peer", local_app_data={"secret-id": secret.id}
+    )
+    database_relation = scenario.Relation(
+        "database", "postgresql_client", remote_app_data={"endpoints": endpoint}
+    )
+    # SCENARIO-NOTE: It would be nice if the charm test didn't need to know the
+    # layout, since the lib is hiding all of that - but maybe that's really an
+    # issue with how the lib hides everything.
+    db_secret = scenario.Secret(
+        "1235",
+        {0: {"username": user, "password": password}},
+        label=f"database.{database_relation.relation_id}.user.secret",
+    )
+    state = scenario.State(
+        leader=True,
+        relations=[peer_relation, database_relation],
+        containers=[background_container],
+        storage=[storage],
+        secrets=[secret, db_secret],
+        config={"main-domain": domain, "media-uri": media},
+    )
+
+    ## Act.
+    out = ctx.run(background_container.pebble_ready_event, state)
+
+    ## Assert.
+    cmd = "python3 manage.py runstator"
+    assert_running_state(
+        ctx,
+        out,
+        out.containers[0],
+        user,
+        password,
+        endpoint,
+        secret.contents[1]["takahe-secret-key"],
+        domain,
+        media,
+        "Takahē background service",
+        cmd,
+        background_container.name,
+    )
+    assert out.workload_version == version
+
+
+def assert_running_state(
+    ctx,
+    out,
+    container,
+    db_user,
+    db_password,
+    db_host,
+    secret_key,
+    domain,
+    media,
+    summary,
+    cmd,
+    service_name,
+):
+    takahē_env = {
+        "TAKAHE_DATABASE_SERVER": f"postgresql://{db_user}:{db_password}@{db_host}/takahē?connect_timeout=10",
+        "TAKAHE_SECRET_KEY": secret_key,
+        "TAKAHE_MEDIA_BACKEND": media,
+        "TAKAHE_MAIN_DOMAIN": domain,
+        "TAKAHE_EMAIL_FROM": f"takahē@{domain}",
+        "TAKAHE_AUTO_ADMIN_EMAIL": f"takahē@{domain}",
+        "TAKAKE_USE_PROXY_HEADERS": "True",
+        "TAKAHE_MEDIA_ROOT": str(out.storage[0].get_filesystem(ctx)),
+        "TAKAHE_MEDIA_URL": "https://example.com/",
+    }
+    # SCENARIO-NOTE: it would be nice if there was a cleaner way to create the
+    # plan object.
+    raw_plan = yaml.safe_dump(
+        {
+            "summary": summary,
+            "description": "Service for Takahē ActivityPub instance",
+            "services": {
+                service_name: {
+                    "override": "replace",
+                    "summary": summary,
+                    "command": cmd,
+                    "startup": "enabled",
+                    "environment": takahē_env,
+                },
+            },
+        }
+    )
+    expected_plan = ops.pebble.Plan(raw_plan)
```

Comment:
@benhoyt This is more an ops/Pebble question than a Scenario one: how do we expect charmers to create plans in tests? I assume we never expect them to be created outside of tests, because charms deal with layers and Pebble makes them into plans.

I could write this as a YAML string and create the `Plan` from that. Is that the intention?

Reply:
We should probably add `Plan.from_dict`, or just make `Plan.__init__` take `str | PlanDict` similar to how `Layer.__init__` does it. In actual charms I think I've seen most people deal with `Layer`s rather than `Plan`s (`Layer` has a from-dict initialiser and and also `__eq__`).